include/circt/Dialect/MSFT/MSFTConstructs.td

```diff
@@ -67,3 +67,48 @@ def ChannelOp: MSFTOp<"constructs.channel",
     $input $clk $sym_name `(` $defaultStages `)` attr-dict `:` type($input)
   }];
 }
+
+// Linear, pipelineable datapath.
+def LinearOp : MSFTModuleOpBase<"hlc.linear", [
```

Comment:
I had expected this to not be a module-like construct. I expected it to be a block of logic inside of another module. See SystolicArray. I imagine the PyCDE interface would be a `with` block or decorated function to allow both RTL-level logic and pipelines to be written side-by-side. (See `constructs.SystolicArray`.) This model more closely resembles that.

Why did you choose to do it this way? Either way, I all of high level constructs should be the same style.

Reply:
The design choice was mainly motivated by that i see these high-level constructs to (eventually) also define dynamic-instance-defined values, e.g.:
```mlir
msft.hlc.linear @M(%in : i32, %clk: i1 {msft.clock}) -> ()
 attributes {
   msft.dynamic_defs = {@latency : i32 }
 } {
   ...
}
```
Which could be referenced by other ops, e.g.:
```mlir
msft.module @foo(...) {
    msft.instance @an_M @M(...)
    %latency = #hw.param.value<#hw.param.decl.ref<@an_M::@latency>>
    ...
}
```

That eventually, during lowering, resolves to a value. In this context, it felt natural to have the construct as a symbol-defining, module-like operation which may be instantiated at multiple sites in different configurations.
Now, this method of having an `msft.dynamic_defs` attribute could (should) also apply to a `MSFTModule` operation, so nothing prevents an embedded version of this op to just push those things to its parent `MSFTModule` if we go that route. LMKWYT.

Reply:
edit: given some thought, i think i'd prefer an embedded version as well. Still interested in hearing your thoughts in relation to parametrization/lowering-defined definitions and if that's more suitable to symbol-defining ops.

Reply:
> The design choice was mainly motivated by that i see these high-level constructs to (eventually) also define dynamic-instance-defined values, e.g.:

It's not just the dynamic-instance case. Even if the pipeline schedule is computed statically, the "calling" module should eventually know the latency and not rely on dynamic scheduling.

> ```mlir
> msft.module @foo(...) {
>     msft.instance @an_M @M(...)
>     %latency = #hw.param.value<#hw.param.decl.ref<@an_M::@latency>>
>     ...
> }
> ```

Is that syntax valid?


This is a good idea, but I'm not sure it's the way we want to go. The designer may want to make different decisions based on the value (in Python). I was thinking something like Futures, but haven't really thought it through.

Reply:
> It's not just the dynamic-instance case. Even if the pipeline schedule is computed statically, the "calling" module should eventually know the latency and not rely on dynamic scheduling.

makes sense - regardless, I think the philosophy of the proposed parametrization scheme still applies - that a module defines parameters which other modules can reference. Those parameters may be resolved at the module level or the dynamic instance level.

> Is that syntax valid?

Dunno - just IR jamming. But I am confident something similar could be made to work.

Reply:
> Having not thought this through, i could see how such a future system could be based on the same style of symbol defines, just like the current hardware parametrization scheme is.
> 
> ```mlir
> msft.module @foo(...) {
>     msft.instance @an_M @M(...)
>     %future= #msft.future<#hw.param.decl.ref<@an_M::@latency>>
>     ...
>     // Things that will run once the future returns.
>     msft.if %future == ... {
>     } else { ... }
> }
> ```

Yeah, that's a level of complexity which I'd rather handle in Python. Having to encode Python logic in IR basically turns PyCDE into an HDL. So I'd rather have the `... {} else {}` be a function pointer. I do think it's reasonable to declare the `future` and list of `future`s which should be passed to the function call in the IR.